Check that built-in callable types validate their output type is `Sized` (in new solver)

compiler-errors opened the pull request:
Working on parity with old solver. Putting this up for consideration, it's not *really* needed or anything just yet. Maybe it's better to approach this from another direction (like always checking the item bounds when calling `consider_assumption`? we may need that for coinduction to be sound though?)

This basically implements #100096 for the new solver.

compiler-errors committed:
Check that built-in callable types validate their output type is `Sized` (in new solver)

lcnr commented on compiler/rustc_trait_selection/src/solve/project_goals.rs:
```diff
@@ -329,22 +330,29 @@ impl<'tcx> assembly::GoalKind<'tcx> for ProjectionPredicate<'tcx> {
                 goal_kind,
             )?
         {
+            // A built-in `Fn` trait needs to check that its output is `Sized`
```
```suggestion
            // A built-in `Fn` impl needs to check that its output is `Sized`
```

lcnr commented on compiler/rustc_trait_selection/src/solve/trait_goals.rs:
```diff
@@ -173,20 +177,27 @@ impl<'tcx> assembly::GoalKind<'tcx> for TraitPredicate<'tcx> {
                 goal_kind,
             )?
         {
+            // A built-in `Fn` trait needs to check that its output is `Sized`
```
```suggestion
            // A built-in `Fn` impl needs to check that its output is `Sized`
```

lcnr reviewed (commented):
Am in favor of landing this PR.

some related thoughts wrt naming n stuff: I am actually not too happy about the name `consider_assumption` given how it is used.

If we have a builtin impl we're not really considering an assumption as it's something where we **know** that it holds, we're simply looking at an impl header with no nested goals.

If we were to change `consider_assumption` to also take a list of nested obligations it would be pretty much how candidates in chalk (and probably a-mir-formality) work?

So maybe a name like `fn consider_normal_candidate` (with a comment explaining what *normal* means might be better? :thinking: idk

compiler-errors committed:
Add consider_implied_clause

compiler-errors commented:
I renamed it to `consider_hypothesis` and moved some stuff around. Not super attached to anything, so please let me know if you think things should be renamed or moved.

@rustbot ready

lcnr commented on compiler/rustc_trait_selection/src/solve/assembly.rs:
```diff
@@ -90,16 +90,20 @@ pub(super) trait GoalKind<'tcx>: TypeFoldable<'tcx> + Copy + Eq {
+    // Use a hypothesis, which consists of a "assumption" and some "requirements",
+    // to satisfy a goal. If the requirements hold, then attempt to satisfy our
+    // goal by equating it with the assumption.
+    fn consider_hypothesis(
```
not too happy with hypothesis https://www.merriam-webster.com/dictionary/hypothesis

> A [hypothesis](https://www.merriam-webster.com/dictionary/hypothesis) is an assumption, an idea that is proposed for the sake of argument so that it can be tested to see if it might be true.
> 
> In the scientific method, the hypothesis is constructed before any applicable research has been done, apart from a basic background review. You ask a question, read up on what has been studied before, and then form a hypothesis.

we know that the `requirements -> assumption` is correct. Looking at chalk, they use `clause` for this. I don't think we should do that in rustc because we already have a `Clause` which (at least for now) is something slightly different.

Other ideas:
- `consider_(logical)_statement` (meh)
- `consider_implication` (this will be how implications in the param env will be implemented (if and once we get them))
- `consider_implied_clause` (changing `assumption` to `clause: ty::Clause`, will actually re)

I personally kinda like `consider_implied_clause` :thinking: It is a clause which is implied by some `requirements`. But yeah, if  you don't like `consider_implied_clause` we can just go with whatever.

r=me after nit

compiler-errors replied:
Yeah, consider_implied_clause is a totally fine name!

compiler-errors commented:
@bors r=lcnr

bors commented:
:pushpin: Commit 6402c98621ab620a16adac072abc10779cf60573 has been approved by `lcnr`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=lcnr 6402c98621ab620a16adac072abc10779cf60573 -->
<!-- homu: {"type":"Approved","sha":"6402c98621ab620a16adac072abc10779cf60573","approver":"lcnr","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 6402c98621ab620a16adac072abc10779cf60573 with merge fcdbd1c07f0b6c8e7d8bbd727c6ca69a1af8c7e9...
<!-- homu: {"type":"BuildStarted","head_sha":"6402c98621ab620a16adac072abc10779cf60573","merge_sha":"fcdbd1c07f0b6c8e7d8bbd727c6ca69a1af8c7e9"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4214193923/jobs/7315451218)
Approved by: lcnr
Pushing fcdbd1c07f0b6c8e7d8bbd727c6ca69a1af8c7e9 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"lcnr","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4214193923/jobs/7315451218"},"merge_sha":"fcdbd1c07f0b6c8e7d8bbd727c6ca69a1af8c7e9"} -->

rust-timer commented:
Finished benchmarking commit (fcdbd1c07f0b6c8e7d8bbd727c6ca69a1af8c7e9): [comparison URL](https://perf.rust-lang.org/compare.html?start=f77f4d55bdf9d8955d3292f709bd9830c2fdeca5&end=fcdbd1c07f0b6c8e7d8bbd727c6ca69a1af8c7e9&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=f77f4d55bdf9d8955d3292f709bd9830c2fdeca5&end=fcdbd1c07f0b6c8e7d8bbd727c6ca69a1af8c7e9&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=f77f4d55bdf9d8955d3292f709bd9830c2fdeca5&end=fcdbd1c07f0b6c8e7d8bbd727c6ca69a1af8c7e9&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 1.0% | [1.0%, 1.0%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=f77f4d55bdf9d8955d3292f709bd9830c2fdeca5&end=fcdbd1c07f0b6c8e7d8bbd727c6ca69a1af8c7e9&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.
